### PR TITLE
avoid unnecessary live file query and fix double referencing moved file

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

Obsolete file candidates consist of `obsolete_files_` that are no longer referenced by any version in `VersionSet`, and all files in DB directory if `doing_full_scan` is true. By design, `obsolete_files_` is guaranteed to be obsolete and thus no need to be checked against live files list. This PR avoids the unnecessary live file query when not doing full scan.

The optimization depends on the assumption that each physical SST file is represented by an unique `FileMetaData` structure, which is actually broken in existing codebase: A trivial file move will create a new file meta for the same file number. This patch also fix this issue.

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
Sorry for delay, will get to this today or tomorrow.

File: db/version_edit.h [hunk 1]
```diff
@@ -129,6 +129,13 @@ struct FileMetaData {
         init_stats_from_file(false),
         marked_for_compaction(false) {}
 
+  void Ref() { ++refs; }
+
+  bool Unref() {
+    --refs;
```
Comment (Connor1996) [inline db/version_edit.h:135]: How about making `ref` be private?

File: db/version_builder.cc [hunk 2]
```diff
@@ -514,10 +515,26 @@ class VersionBuilder::Rep {
...
+      deleted_base_files_[f->fd.GetNumber()] = f;
+    } else {
+      vstorage->AddFile(level, f, info_log_);
+    }
+  }
+
+  void MaybeAddFile(VersionStorageInfo* vstorage, int level, FileMetaData* f,
+                    uint64_t base_max_file_number) {
+    if (f->fd.GetNumber() <= base_max_file_number) {
+      auto it = deleted_base_files_.find(f->fd.GetNumber());
```
Comment (Connor1996) [inline db/version_builder.cc:532]: why not just use `deleted_base_files_` to decide if it is a trivial move.

Comment (tabokie) [reply db/version_builder.cc:532]: To make a stronger assertion below. If it's not included in `deleted_base_files_` but have a small file number, it means we are creating a new file with an old name.

File: db/version_builder.cc [hunk 3]
```diff
@@ -514,10 +515,26 @@ class VersionBuilder::Rep {
...
+      assert(it != deleted_base_files_.end());
+      vstorage->AddFile(level, it->second, info_log_);
```
Comment (yiwu-arbug) [inline db/version_builder.cc:534]: Do we need to bump ref count for the added file?

**Review by yiwu-arbug — COMMENTED** [review]
Add more comments overall?

File: db/version_builder.cc (hunk 3 above)
Comment (tabokie) [reply db/version_builder.cc:534]: `AddFile` will do that.

File: db/version_builder.cc [hunk 4]
```diff
@@ -514,10 +515,26 @@ class VersionBuilder::Rep {
...
+    } else {
+      vstorage->AddFile(level, f, info_log_);
+    }
+  }
+
+  void MaybeAddFile(VersionStorageInfo* vstorage, int level, FileMetaData* f,
+                    uint64_t base_max_file_number) {
+    if (f->fd.GetNumber() <= base_max_file_number) {
+      auto it = deleted_base_files_.find(f->fd.GetNumber());
+      assert(it != deleted_base_files_.end());
```
Comment (tabokie) [inline db/version_builder.cc:533]: @yiwu-arbug @Connor1996 Do you think we need to set a background error when this assertion fails? Violation of this is almost equivalent to data loss.

Comment (Connor1996) [reply db/version_builder.cc:533]: `assert` is disabled for non-debug, so better set a background error if possible

Comment (tabokie) [reply db/version_builder.cc:533]: Just notice this method is allowed to fail. Change to return status instead.

File: db/db_impl/db_impl_files.cc [hunk 5]
```diff
@@ -234,6 +233,9 @@ void DBImpl::FindObsoleteFiles(JobContext* job_context, bool force,
   job_context->log_recycle_files.assign(log_recycle_files_.begin(),
                                         log_recycle_files_.end());
   if (job_context->HaveSomethingToDelete()) {
+    if (doing_the_full_scan) {
+      versions_->AddLiveFiles(&job_context->sst_live);
```
[the remark this answers is not in the document]
Comment (tabokie) [reply db/db_impl/db_impl_files.cc:237]: I think look up an empty map there won't hurt. But I will change the name of this field to something less protruding, it doesn't always contain the full sst live files anymore.

File: db/version_edit.h (hunk 1 above)
Comment (yiwu-arbug) [reply db/version_edit.h:135]: +1

**Comment by yiwu-arbug** [comment]
Looking good overall. Let's send to upstream and see?

**Comment by yiwu-arbug** [comment]
duplicated with #250

**Pull request closed without merging**